Indexing speedups

KCMertens opened the pull request:
Several low hanging fruit improvements:
- better xpath caching in saxon-related code. Turns out setting up the execution environment and loading variables is slow, so cache everything *except* the context step.
- Rewrite recursive directory walking to use walkFileTree, which seems to perform immensely better in some scenarios, namely under windows when using remote filesystem mounts (sshfs/smb shares). It seems like it's something to do with repeatedly calling isFile etc() vs doing it in one batch under the hood. Also remove the file order guarantee in some places where it's not required (such as gathering files to index), as the comparator for sorting files repeatedly calls the slow isFile/isDirectory etc, which don't seem to be cached under the hood, causing millions of network requests when the mount is remote. This only seems to be an issue under Windows.
- The insert/upsert check repeatedly created an IndexReader to check if a new document's PID was in use. Since the IndexWriter seems to be a singleton for a given index, we can just keep track of used PIDS directly. This saved something like 30% cpu time across the entire process
- Speed up and simplify the code for tracking exact XML element start/end character offsets within input documents.   
Turns out the XML StAX parser stack reports these directly with events, whereas SAX does not. Since Saxon can parse documents using StAX as well, use that, and leave the offset tracking to the parser instead of manually keeping track of newlines etc. There's two caveats;   

> [!NOTE]  
> Replaced the StAX parser with `woodstock` instead of the builtin `Xerces` due to prevalent bugs with character position reporting that were getting very hard to work around. **This makes the described workarounds obsolute. I've removed them.**
<details><summary>Original</summary>
<p>
the start_element event has the char offset position reported at the end of the element opening tag, whereas we need the start.   To solve this, we still track the brackets manually, and walk back from the reported position. This works because in legal XML, the `<` character is illegal in its entirety in an element opener/closer. So e.g. `<element att="<">`, which would break this assumption, wouldn't even parse to begin with.  
```
  <element>...</element>
          ^---------- reported position
  ^------------------ required position
```
In some cases, the `end_element` event reports a position 1 character after the closer. From my investigations this seems to be a bug in the Xerces implementation? In any case, we also track the `>` characters and correct for this.
</p>
</details> 

- Then there's one more involved change.
Update the way we interact with Saxon by using XdmValue instead of NodeInfo objects in many places. The `XdmValue` represents a Saxon "some value" i.e. basically anything that can be returned from an xpath, whereas the NodeInfo is specifically an XML DOM node (element/document/attribute/comment/cdata I think). 
This both saves some repeated wrapping/unwrapping (with pretty extensive internal sanity checks in Saxon), as well as allowing many xpaths to return non-document things, such as (collections of) strings. 
This opens up patterns like this:
```yaml
# for a given part of speech that looks like e.g. 'ADP(type=pre)+PD(type=d-p,subtype=art,position=prenom)'
 - name: pos 
      displayName: Part of Speech
      basePath: "tokenize(@pos, '\\+')" # ('ADP(type=pre)', 'PD(type=d-p,subtype=art,position=prenom)'
      valuePath: "tokenize(., '\\(')[1]" # ('ADP', 'PD') -- called on all results of basePath. NOTE: use tokenize as substring-before returns '' if the char '(' is not present
      subannotations:
        - forEachPath: "tokenize(substring-before(substring-after(., '('), ')'), ',')" # ('type=pre', 'type=d-p', 'subtype=art', 'position=prenom')
          namePath: substring-before(., '=')  # e.g. type
          valuePath: "tokenize(substring-after(., '='), '\\|')"  # e.g. pre
```
Basically it just opens up using partially pre-processed values instead of having to repeatedly find/split/tokenize the same value to put it into similar annotations.

PrinsINT committed:
Refactor and speed up stripAccents

- remove unused parameter removeEmSpace (the method name also implies that it lies beyond its functionality)
- Loop over chars instead of regex

KCMertens committed:
More aggressively cache xpaths to speed up indexing

Even though the expressions were cached, the namespaces and variables associated with the xpath were not.
This represented a significant overhead.

KCMertens committed:
Featureflag/speedup: defer segment merges during indexing

Adds the feature flag `deferSegmentMergesDuringIndexing`.
Enabling the feature (a string that parses to a boolean true value) disables all segment merges during indexing
then starts a repeated merge using the standard TieredMergePolicy after indexing is complete, until no more merges are needed.
This should speed up initial indexing speed, at the cost of a wait after indexing to complete the merges.
The net result should still be net positive, as unnecessary merges during indexing are avoided.

KCMertens committed:
Speedup: gather files using walkFileTree

regular file iterating is super slow on networked storage. Sorting files makes it worse (repetitive networks calls for .isFile() and .isDirectory()).

This still needs some further refinement however. We have a couple methods to iterate across files we use interchangably, and we should port them all to the fast path.

KCMertens committed:
Speedup: retrieve used pids only once during indexing

Does mean we need to keep our own list in check, and it would break if there's concurrent IndexWriters.
However BlackLab always has a singleton IndexWriter for a given index.

KCMertens committed:
Allow more advanced xpaths in annotations and metadata

Non-terminal xpaths (i.e. those that can serve as input context for subsequent xpaths, such as containerPath, or forEachPath) had to return a tangible node in the document (attribute, text or element).
This prevents some optimizations, like splitting a string right in the xpath, and indexing all sections under different subannotations.
Refactor the DocindexerSaxon to work directly with the xpath result primitives, instead of the underlying document nodes to allow this.
For some xpaths, we still need a tangible document node, because BlackLab needs to know the offsets in the document (for highlighting purposes), but this is no different than before, and you'll get the same exception if your xpath doesn't result in a node where BlackLab expects one.

In practice, this means you can now use xpaths like `tokenize(., 'some-split-char')` in the following locations where it was previously disallowed:
- (sub)annotation.basePath
- (sub)annotation.forEachPath
- metadatafield.forEachPath

KCMertens committed:
Faster charpostracking implementation? Vastly simpler at least!

Undo speedup.
Improve and speed up element offset tracking during XML parse.
cleanup old system.
Fix incorrect (overflowing) charpositions reported for some documents.

KCMertens committed:
Cache canonical paths; retrieval is rather slow

KCMertens committed:
Use iterators in XpathFinder; avoids XdmValue wrapping

This is quite a bit faster than the somewhat more fluid XdmValue approach we were using before, though speed depends on the internal implementation in Saxon, which (currently) seems to have Iterators as the base level, with most functions wrapping the iterators. By skipping the intermediate functions we get a modest speedup.

KCMertens committed:
Fix an off by 1 error for element end positions in long documents

There seems to be an internal bug in Xerces where element end offsets are reported off by 1 after buffer refill (at 8192). Track the closing brackets as well, and fix the offset.
Add some more tests around the specific error.

jan-niestadt commented on engine/src/main/java/nl/inl/blacklab/plugins/IndexSourceType.java:
```diff
@@ -8,7 +8,7 @@
 
-    private static final String PROTOCOL_SEPARATOR = ":";
+    private static final String PROTOCOL_SEPARATOR = "//:"; // avoid matching random ':' or windows pathnames.
 
```
This wasn't quite right. Should handle URIs like '/tmp/bla.txt', 'file:/tmp/bla.txt', 'http://nu.nl/' and 'db:12345'. Fixed in e7dd48f47, added test, including for Windows path

jan-niestadt replied:
Fixed in e7dd48f47, added test, including for Windows path

KCMertens replied:
Ahh, I somehow had file urls in my head as `file://`, good catch!

jan-niestadt replied:
I think `//` is optional for URIs in general. Just tried it and Firefox and Chrome both accept e.g. `https:ivdnt.org` :)

I should probably recognize the slashes if they're there and automatically remove them.

KCMertens replied:
TIL

jan-niestadt committed:
IndexSourceType: deal with colon in path. Test.

jan-niestadt committed:
Synchronize BLIndexWriterProxyLucene.usedPids. Refactor a bit.

jan-niestadt commented on engine/src/main/java/nl/inl/blacklab/index/BLIndexWriterProxyLucene.java:
```diff
@@ -119,4 +146,51 @@ public long updateDocument(Term term, BLInputDocument document) throws IOExcepti
+     * @return true if it's a new pid, false if it was already present
+     */
+    private synchronized boolean addToPids(String pid) {
+        if (usedPids == null) {
```
We should synchronize on usedPids, added that in 03b743fe4

KCMertens replied:
Agree

jan-niestadt commented on engine/src/main/java/nl/inl/blacklab/indexers/config/saxon/XPathFinder.java:
```diff
@@ -41,38 +47,49 @@ public class XPathFinder {
-        xPath = SaxonHelper.newXPathFactory();
-        xPath.setCaching(true); // cache xpaths
+    private final LoadingCache<Map<String, String>, XPathCompiler> compilerCache = Caffeine.newBuilder()
+        .maximumSize(50) // should be large enough for most use cases?
```
Should be static, or it won't cache anything? But then we can't access vars, which currently only contains fromInputFile, which is different for every indexed file, so we still wouldn't benefit from caching?

jan-niestadt committed:
Various cleanups, fixes, refactors.

- Code style. Convert tabs to spaces.
- XPathFinder.compiledXpaths cannot be static, values depend on vars.
- Fix SaxonDocumentWithElementOffsets.openElementStack.
  IntelliJ indicated that the initializer was redundant because the constructor sets openElementStack to null. But this is after it has been used. Moved the initializer into the constructor to inside the constructor to prevent this confusion.
- Remove obsolete comment.
- IndexSourceType: skip optional //

jan-niestadt commented on engine/src/main/java/nl/inl/blacklab/indexers/config/saxon/XPathFinder.java:
```diff
@@ -41,38 +47,49 @@ public class XPathFinder {
 
-        // xml namespace is implicit
-        xPath.declareNamespace(NAMESPACE_XML_PREFIX, NAMESPACE_XML_URI);
+    public XPathFinder(Map<String, String> namespaces, Map<String, String> vars) {
```
Can we safely cache compiled XPaths? XpathSelector depends on vars, which change for each document indexed. (removed static in 1a04c8efb)

KCMertens replied:
Hmm, that's unfortunate. The slow part is the `XpathExecutable.load()` function. As I understand it it creates various internal structures for holding the declared variables, the context value, etc. 

Thinking through it, there's some steps to compiling xpath, each of which can/should be cached.
1. set up the `XPathCompiler` , for this we need: 
    - the namespaces and their values
    - variable names (though not their values)
    
    **This is fairly slow** 
    By removing the `static` from the cache, this step is effectively no longer cached now, since we construct a new `XPathFinder` per document, and a new `XPathCompiler` per `XPathFinder`. We're now left with a useless cache for every document, the cache just holding 1 factory instance. 
    We should see if we could make it so that the cache primary key is the combination of namespaces + variable declarations _(not variable values, those don't matter at this point)_, then we could share these factories across index threads and documents again.

2. compile xpath to obtain `XPathExecutable`. **these are what `XPathCompiler` will cache internally**, so no special behavior required for us here.

3. call `load()` on the `XPathExecutable` to obtain `XPathSelector`

   **this is slow, and Saxon doesn't cache it** .  
   These objects are reusable across documents and variable values. 
   Since they're tied to the Compiler instance, we could probably cache a ThreadLocal keyed on the original XPathFactory to store these. This way we make sure every index thread will get its own instance of the xpath for the given set of variables and namespaces.  Essentially `static Cache<XPathFactory, ThreadLocal<XpathSelector>>`. 
 
4. set the context node and variable values on the xpath. This is fairly fast and can be done on obtain

jan-niestadt commented on engine/src/main/java/nl/inl/blacklab/indexers/config/saxon/SaxonDocumentWithElementOffsets.java:
```diff
@@ -0,0 +1,196 @@
+	private LongArrayList closeBracketPositions = new LongArrayList();
+	/** Contains the start/end pairs of currently open elements. */
+	private ObjectList<LongLongPair> openElementStack = new ObjectArrayList<>();
+
```
IntelliJ indicated that the openElementStack initializer was redundant, so I removed it, which broke things. I've moved the initializer into the constructor to avoid this confusion.

jan-niestadt reviewed (commented):
Some really nice optimizations and improvements! I did notice a few potential issues, could you have a look and see if you agree?

BTW I replaced a bunch of tabs to spaces throughout the project for consistency, but fortunately you can hide whitespace changes with the gear icon under "Files changed".

KCMertens committed:
Fix XPath caching

Thinking through it, there's some steps to compiling xpath, each of which can/should be cached.
1. set up the `XPathCompiler` , for this we need:
    - the namespaces and their values
    - variable names (though not their values)

    **This is fairly slow**
    By removing the `static` from the cache, this step is effectively no longer cached now, since we construct a new `XPathFinder` per document, and a new `XPathCompiler` per `XPathFinder`. We're now left with a useless cache for every document, the cache just holding 1 factory instance.
    We should see if we could make it so that the cache primary key is the combination of namespaces + variable declarations _(not variable values, those don't matter at this point)_, then we could share these factories across index threads and documents again.

2. compile xpath to obtain `XPathExecutable`. **these are what `XPathCompiler` will cache internally**, so no special behavior required for us here.

3. call `load()` on the `XPathExecutable` to obtain `XPathSelector`

   **this is slow, and Saxon doesn't cache it** .
   These objects are reusable across documents and variable values.
   Since they're tied to the Compiler instance, we could probably cache a ThreadLocal keyed on the original XPathFactory to store these. This way we make sure every index thread will get its own instance of the xpath for the given set of variables and namespaces.  Essentially `static Cache<XPathFactory, ThreadLocal<XpathSelector>>`.

4. set the context node and variable values on the xpath. This is fairly fast and can be done on obtain

Fix regression; declaring the default 'xml' namespace shouldn't enable namespace-sensitivity

KCMertens committed:
Replace sun/oracle stax parser with Woodstox; fixes random character offset bugs

KCMertens committed:
Reduce memory usage of element offset tracking code

jan-niestadt committed:
Re-fix to correctly remove linked documents if too old or cache too large.
